soroban-env-host/src/host/diagnostics_helper.rs

```diff
@@ -0,0 +1,105 @@
+use soroban_env_common::{
+    xdr::{ContractEventType, Hash, ScVal, ScVec},
+    ConversionError, ScValObjRef, Symbol, TryFromVal, TryIntoVal, VecObject,
+};
+
+use crate::events::{InternalContractEvent, InternalEvent};
+use crate::{budget::AsBudget, Host, HostError, RawVal};
+
+/// None of these functions are metered, which is why they're behind the is_debug check
+impl Host {
+    fn hash_to_scval(&self, hash: Hash) -> Result<ScVal, HostError> {
+        Ok(ScVal::Bytes(self.scbytes_from_hash(&hash)?))
+    }
+
+    fn record_system_debug_contract_event(
+        &self,
+        type_: ContractEventType,
+        topics: VecObject,
+        data: RawVal,
+    ) -> Result<(), HostError> {
+        let ce = InternalContractEvent {
+            type_,
+            contract_id: None,
+            topics,
+            data,
+        };
+        self.get_events_mut(|events| {
+            Ok(events.record(InternalEvent::StructuredDebug(ce), self.as_budget()))
+        })?
+    }
+
+    // Emits an event with topic = ["fn_call", contract_id, function_name] and
+    // data = [arg1, args2, ...]
+    pub fn fn_call_diagnostics(
+        &self,
+        contract_id: &Hash,
+        func: &Symbol,
+        args: &[RawVal],
+    ) -> Result<(), HostError> {
+        if !self.is_debug() {
+            return Ok(());
+        }
+
+        self.as_budget().with_free_budget(|| {
+            let mut topics: Vec<ScVal> = Vec::new();
+            topics.push(ScVal::Symbol(crate::xdr::ScSymbol(
```

Comment:
particularly for these small topic symbols, they can/should be `const SymbolSmall`s.

Reply:
(At some point I want to make an explicit local cache of host objects and rawvals we compute once and then reuse repeatedly, like the current contract ID and various symbols and such, so if it's challenging to make work here then don't bother for now, but I think it should work locally here)

Reply:
This now uses `SymbolSmall`.